### PR TITLE
 refactor: use async/await instead of callbacks

### PR DESCRIPTION
**Pull request by jacobheun**
BREAKING CHANGE: The api now uses async/await instead of callbacks.

* Updates all dependencies to their latest async versions
* Adds node 12 to travis
* Adds webworkers to the test suite
* Updates the readme api, and cleans up some links
* Removes pre push test, but keeps lint

### CODE REVIEW

File: src/cms.js [hunk 1]
```diff
@@ -40,44 +36,29 @@ class CMS {
    *
    * @param {string} name - The local key name.
    * @param {Buffer} plain - The data to encrypt.
-   * @param {function(Error, Buffer)} callback
    * @returns {undefined}
    */
-  encrypt (name, plain, callback) {
+  async encrypt (name, plain) {
     const self = this
```
Comment (achingbrain) [inline src/cms.js]: Probably don't need to `self = this` any more now that scope is simpler.

File: package.json [hunk 2]
```diff
@@ -42,26 +39,24 @@
...
-    "interface-datastore": "~0.6.0",
-    "libp2p-crypto": "~0.16.1",
+    "err-code": "^2.0.0",
+    "interface-datastore": "^0.7.0",
+    "libp2p-crypto": "^0.17.0",
     "merge-options": "^1.0.1",
-    "node-forge": "~0.7.6",
-    "pull-stream": "^3.6.9",
+    "node-forge": "^0.8.5",
+    "promisify-es6": "^1.0.3",
```
Comment (achingbrain) [inline package.json]: Should this be a dev dep? It's added by this PR but only used in `test/node.js` that I can see.

**Comment by jacobheun** [comment]
👍 nits fixed

File: package.json [hunk 3]
```diff
@@ -12,16 +12,13 @@
     "test:browser": "aegir test -t browser",
     "release": "aegir release",
     "release-minor": "aegir release --type minor",
-    "release-major": "aegir release --type major",
-    "coverage": "aegir coverage",
-    "coverage-publish": "aegir coverage publish"
+    "release-major": "aegir release --type major"
```
Comment (alanshaw) [inline package.json]: Even though aegir doesn't support it I've kept a `coverage` script as `nyc --reporter=text --reporter=lcov npm run test:node` because `npm run coverage` is way easier to remember.

File: src/util.js [hunk 4]
```diff
@@ -67,5 +66,24 @@ exports.certificateForKey = (key, privateKey, callback) => {
...
+ *
+ * `asyncCompare` is an async function that must
+ * resolve to either `true` or `false`.
+ *
+ * @param {Array} array
+ * @param {function(*)} asyncCompare An async function that returns a boolean
+ */
+async function findAsync (array, asyncCompare) {
+  const promises = array.map(asyncCompare)
+  const results = await Promise.all(promises)
```
Comment (alanshaw) [inline src/util.js]: We have to wait for all these promises to resolve before we can find the value. This is a change in behaviour from `async/detect` which executes in parallel but will call the callback _as soon as _ one returns true.

...we can leave this for a separate PR.

File: src/keychain.js [hunk 5]
```diff
@@ -175,223 +172,195 @@ class Keychain {
    * @param {string} name - The local key name; cannot already exist.
    * @param {string} type - One of the key types; 'rsa'.
    * @param {int} size - The key size in bits.
-   * @param {function(Error, KeyInfo)} callback
    * @returns {undefined}
```
Comment (vasco-santos) [inline src/keychain.js]: Should we also specify in jsdocs the content of `KeyInfo`?

```suggestion
    * @returns {KeyInfo}
```

File: src/keychain.js [hunk 6]
```diff
@@ -175,223 +172,195 @@ class Keychain {
...
+    }
+
+    return keyInfo
   }
 
   /**
    * List all the keys.
    *
-   * @param {function(Error, KeyInfo[])} callback
    * @returns {undefined}
```
Comment (vasco-santos) [inline src/keychain.js]:
```suggestion
    * @returns {KeyInfo[]}
```

File: src/keychain.js [hunk 7]
```diff
@@ -175,223 +172,195 @@ class Keychain {
...
+
+    return info
   }
 
   /**
    * Find a key by it's id.
    *
    * @param {string} id - The universally unique key identifier.
-   * @param {function(Error, KeyInfo)} callback
    * @returns {undefined}
```
Comment (vasco-santos) [inline src/keychain.js]:
```suggestion
    * @returns {KeyInfo}
```

File: src/keychain.js [hunk 8]
```diff
@@ -175,223 +172,195 @@ class Keychain {
...
+      return throwDelayed(err)
+    }
   }
 
   /**
    * Find a key by it's name.
    *
    * @param {string} name - The local key name.
-   * @param {function(Error, KeyInfo)} callback
    * @returns {undefined}
```
Comment (vasco-santos) [inline src/keychain.js]:
```suggestion
    * @returns {KeyInfo}
```

File: src/keychain.js [hunk 9]
```diff
@@ -175,223 +172,195 @@ class Keychain {
...
+      return throwDelayed(errcode(new Error(`Key '${name}' does not exist. ${err.message}`), 'ERR_KEY_NOT_FOUND'))
+    }
   }
 
   /**
    * Remove an existing key.
    *
    * @param {string} name - The local key name; must already exist.
-   * @param {function(Error, KeyInfo)} callback
    * @returns {undefined}
```
Comment (vasco-santos) [inline src/keychain.js]:
```suggestion
    * @returns {KeyInfo}
```

File: src/keychain.js [hunk 10]
```diff
@@ -175,223 +172,195 @@ class Keychain {
...
+    return keyInfo
   }
 
   /**
    * Rename a key
    *
    * @param {string} oldName - The old local key name; must already exist.
    * @param {string} newName - The new local key name; must not already exist.
-   * @param {function(Error, KeyInfo)} callback
    * @returns {undefined}
```
Comment (vasco-santos) [inline src/keychain.js]:
```suggestion
    * @returns {KeyInfo}
```

File: src/keychain.js [hunk 11]
```diff
@@ -175,223 +172,195 @@ class Keychain {
...
-            if (err) return _error(callback, err)
-            callback(null, keyInfo)
-          })
-        })
-      })
-    })
+    let res = await this.store.get(oldDsname)
+    const pem = res.toString()
+
+    const exists = await self.store.has(newDsname)
```
Comment (vasco-santos) [inline src/keychain.js]: What about doing the new validating first, then the get of the old one?

File: src/keychain.js [hunk 12]
```diff
@@ -175,223 +172,195 @@ class Keychain {
...
+    return keyInfo
   }
 
   /**
    * Export an existing key as a PEM encrypted PKCS #8 string
    *
    * @param {string} name - The local key name; must already exist.
    * @param {string} password - The password
-   * @param {function(Error, string)} callback
    * @returns {undefined}
```
Comment (vasco-santos) [inline src/keychain.js]:
```suggestion
    * @returns {string}
```

File: src/keychain.js [hunk 13]
```diff
@@ -400,99 +369,96 @@ class Keychain {
    * @param {string} name - The local key name; must not already exist.
    * @param {string} pem - The PEM encoded PKCS #8 string
    * @param {string} password - The password.
-   * @param {function(Error, KeyInfo)} callback
    * @returns {undefined}
```
Comment (vasco-santos) [inline src/keychain.js]:
```suggestion
    * @returns {KeyInfo}
```

File: src/keychain.js [hunk 14]
```diff
@@ -400,99 +369,96 @@ class Keychain {
...
-      })
-    })
+    const exists = await self.store.has(dsname)
+    // if (err) return throwDelayed(callback, err)
+    if (exists) return throwDelayed(errcode(new Error(`Key '${name}' already exists`), 'ERR_KEY_ALREADY_EXISTS'))
+
+    const kid = await privateKey.id()
+    // if (err) return throwDelayed(callback, err)
+    const pem = await privateKey.export(this._())
+    // if (err) return throwDelayed(callback, err)
```
Comment (vasco-santos) [inline src/keychain.js]: Should we remove all the above comments?

We should probably add a `try...catch` from the `const kid` line until the commit throwing using the throwDelayed if we catch any error?

File: src/keychain.js [hunk 15]
```diff
@@ -400,99 +369,96 @@ class Keychain {
...
+    await batch.commit()
+    return keyInfo
   }
 
   /**
    * Gets the private key as PEM encoded PKCS #8 string.
    *
    * @param {string} name
-   * @param {function(Error, string)} callback
    * @returns {undefined}
```
Comment (vasco-santos) [inline src/keychain.js]:
```suggestion
    * @returns {string}
```

File: src/keychain.js (hunk 14 above)
Comment (jacobheun) [reply src/keychain.js]: Done, I also wrapped renameKey to make sure we're delaying there as well.

File: src/keychain.js (hunk 11 above)
Comment (jacobheun) [reply src/keychain.js]: Makes sense, done!
